utils/sample_data_generators.py

```diff
@@ -0,0 +1,25 @@
+import math
+
+
+"""
+Generates number with one too many digits than specified, based on e mathematical constant.
+Example: digits=4 will return 135.91
+"""
+
+
+def generate_too_many_digits(digits):
+    assert isinstance(digits, int)
+    assert digits > 1
+    return round((math.e / 2 * 10 ** (digits - 2)), 2)
+
+
+"""
+Generates number with one too many decimal places than specified, based on e mathematical constant.
+Example: decimal_places=2 will return 1.359
```

Comment:
Thought it made sense, since that way you just pass the limit from model constants file for this field. That way, just like function name suggests, it will create value exceeding this limit with marginal difference.

I dunno, it's just me, but personally I find it cleaner than hardcoding these differences like in example above.
I would agree on changing decimal_places parameter from limit value to actual generated value.
Then for testing purposes I would parametrise only this parameter (no difference value to append further), since currently, the way my tests are structured, I use this function only for fail tests.

Reply:
In tests we should hardcode some differences to show them that they are differences. I think, that the best way would be just to create function which are creating correct data and in tests change it to malformed for test needs.

Reply:
I hardcoded correct data in my unit tests for models and serializers, rather than deriving them in any way (in my head it's as if user just types it in from some interface). The incorrect values are currently derived from validators' values, so I can get data which exceeds any limit by smallest possible margin. So in my code it's
```
WORK_HOURS_DATA_EXCEEDING_DIGITS_NUMBER = generate_too_many_digits(ReportModelConstants.MAX_DIGITS.value)
    WORK_HOURS_DATA_EXCEEDING_DECIMAL_PLACES = generate_too_many_decimal_places(
                                                    ReportModelConstants.DECIMAL_PLACES.value)
```

I mean, I obviously agree that reusability is a great feature when designing functions. I could even fuse these two functions and expand them to something like
`def generate_decimal(digits=3, decimal_places=2, math_constant=math.e)`, if it's desired. Then I'd change the unit tests to something similar:
```
MARGINALLY_LARGER_DIGITS_NUMBER = ReportModelConstants.MAX_DIGITS.value + 1

WORK_HOURS_DATA_EXCEEDING_DIGITS_NUMBER = generate_decimal(digits = MARGINALLY_LARGER_DIGITS_NUMBER)
```
However, they were created specifically for supplying data for Report related tests. And since that's the only place in the entire app where decimals are used, it will most certainly be their only usage.
Just curious about your general opinion on this idea, really.

Reply:
@maciejSamerdak 
>Then I'd change the unit tests to something similar:
> 
> ```
> MARGINALLY_LARGER_DIGITS_NUMBER = ReportModelConstants.MAX_DIGITS.value + 1
> 
> WORK_HOURS_DATA_EXCEEDING_DIGITS_NUMBER = generate_decimal(digits = MARGINALLY_LARGER_DIGITS_NUMBER)
> ```
> However, they were created specifically for supplying data for Report related tests. And since that's the only place in the entire app where decimals are used, it will most certainly be their only usage.
> Just curious about your general opinion on this idea, really.

Instead of giving variables like you mentioned above you should do it exactly it tests e.g.:
```python
    def test_report_serializer_work_hours_field_should_not_accept_value_exceeding_set_digits_number(self):
        self.field_should_not_accept_input(field='work_hours', value=ReportModelConstants.MAX_DIGITS.value + 1)

    def test_report_serializer_work_hours_field_should_not_accept_value_exceeding_set_decimal_places_number(self):
        self.field_should_not_accept_input(field='work_hours', value=ReportModelConstants.DECIMAL_PLACES.value + 1)
```

It will reduce code boilerplate and increase tests readability. Please apply this to our tests.

Reply:
Done